File: Python/8-Graphs(BFS-DFS)/Algorithms/1-Depth-FirstSearch(DFS)/Depth-FirstSearch(DFS)Pro.py
"""
Depth-First Search (DFS) Implementation with Six Real-World Applications

DFS is a graph traversal algorithm that explores as far as possible along each branch before backtracking.

Here are six different approaches with real-world project implementations:

1.  **Web Crawler:**
    * **Concept:** DFS can be used to traverse the links of a website. Starting from a root URL, the crawler visits linked pages recursively.
    * **Real-World Application:** Search engine indexing, website link analysis.

2.  **Maze Solving:**
    * **Concept:** Represent a maze as a graph where each cell is a node, and paths between cells are edges. DFS can find a path from the start to the end.
    * **Real-World Application:** Robotics navigation, game development (pathfinding).

3.  **Topological Sort:**
    * **Concept:** For a directed acyclic graph (DAG), DFS can produce a topological ordering of nodes.  A topological sort is an ordering of nodes such that for every directed edge from node A to node B, node A appears before node B in the ordering.
    * **Real-World Application:** Task scheduling, dependency resolution in software projects.

4.  **Cycle Detection:**
    * **Concept:** DFS can detect cycles in a graph. If a visited node is encountered again along the current path, there's a cycle.
    * **Real-World Application:** Detecting circular dependencies in data, detecting loops in linked lists.

5.  **Connected Components:**
    * **Concept:** DFS can identify groups of connected nodes in an undirected graph.  Each group of nodes is a connected component.
    * **Real-World Application:** Network analysis, clustering, image segmentation.

6.  **Sudoku Solver:**
    * **Concept:** DFS can be used to explore possible solutions to a Sudoku puzzle.  The grid can be thought of as a graph, where each empty cell is a node, and possible values are the edges.
    * **Real-World Application:** Artificial Intelligence, Constraint Satisfaction Problems.
"""

from collections import defaultdict
import time

class Graph:
    """
    A graph data structure implemented using an adjacency list.
    """
    def __init__(self):
        """
        Initializes an empty graph.
        """
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        """
        Adds an undirected edge between nodes u and v.
        """
        self.graph[u].append(v)
        self.graph[v].append(u)  # For undirected graph

    def add_directed_edge(self, u, v):
        """
        Adds a directed edge from node u to node v.
        """
        self.graph[u].append(v)

    def get_nodes(self):
        """
        Returns all nodes in the graph
        """
        return list(self.graph.keys())

    def get_neighbors(self, node):
        """
        Returns all neighbors of a given node.
        """
        return self.graph[node]

    def display_graph(self):
        """
        Displays the graph as an adjacency list.
        """
        for node in self.graph:
            print(f"{node}: {self.graph[node]}")

    def dfs(self, start_node, visited=None):
        """
        Performs Depth-First Search (DFS) starting from a given node.

        Args:
            start_node: The node to start the DFS traversal from.
            visited: A set to keep track of visited nodes (optional).

        Returns:
            A list of nodes visited in DFS order.
        """
        if visited is None:
            visited = set()
        visited.add(start_node)
        print(f"Visiting node: {start_node}")  # Track the order of node visits.
        neighbors = self.graph[start_node]
        for neighbor in neighbors:
            if neighbor not in visited:
                self.dfs(neighbor, visited)
        return list(visited)

# 1. Web Crawler
def web_crawler(start_url, max_depth=3):
    """
    Simulates a basic web crawler using DFS.

    Args:
        start_url: The starting URL for the crawl.
        max_depth: The maximum depth to crawl.

    Returns:
        A set of visited URLs.
    """
    visited_urls = set()
    stack = [(start_url, 0)]  # Use a stack for DFS, store (url, depth)

    print(f"\n[Web Crawler] Starting from: {start_url}, Max Depth: {max_depth}")

    while stack:
        current_url, depth = stack.pop()

        if current_url in visited_urls:
            continue
        visited_urls.add(current_url)
        print(f"Crawling: {current_url} (Depth: {depth})")

        if depth >= max_depth:
            continue  # Stop crawling at max depth

        # Simulate fetching links from the current URL
        links = get_links_from_url(current_url)  #  Replace with actual link extraction

        for link in links:
            if link not in visited_urls:
                stack.append((link, depth + 1))  # Push new links onto the stack

    return visited_urls

def get_links_from_url(url):
    """
    Simulates fetching links from a URL (replace with actual implementation).

    Args:
        url: The URL to fetch links from.

    Returns:
        A list of dummy URLs.
    """
    # In a real web crawler, you would use a library like requests and
    # BeautifulSoup to fetch the HTML content and extract links.
    # Here, we just return some dummy links for demonstration.
    if url == "https://www.example.com":
        return ["https://www.example.com/page1", "https://www.example.com/page2", "https://www.example.com/page3"]
    elif url == "https://www.example.com/page1":
        return ["https://www.example.com/page4", "https://www.example.com/page5"]
    elif url == "https://www.example.com/page2":
        return ["https://www.example.com/page6"]
    else:
        return []  # No links for other dummy URLs

# 2. Maze Solving
def solve_maze_dfs(maze, start, end):
    """
    Solves a maze using DFS.

    Args:
        maze: A 2D list representing the maze. 0 is path, 1 is wall.
        start: The starting coordinates (row, col).
        end: The ending coordinates (row, col).

    Returns:
        A list of coordinates representing the path from start to end, or None if no path exists.
    """
    rows = len(maze)
    cols = len(maze[0])
    path = []
    visited = set()

    def is_valid(r, c):
        """
        Checks if a cell is within the maze bounds and is a valid path.
        """
        return 0 <= r < rows and 0 <= c < cols and maze[r][c] == 0

    def dfs(r, c):
        """
        DFS function to explore the maze.
        """
        if (r, c) == end:
            path.append((r, c))
            return True

        if not is_valid(r, c) or (r, c) in visited:
            return False

        visited.add((r, c))
        path.append((r, c))

        # Explore adjacent cells (up, down, left, right)
        if dfs(r - 1, c) or \
                dfs(r + 1, c) or \
                dfs(r, c - 1) or \
                dfs(r, c + 1):
            return True

        path.pop()  # Backtrack: remove the current cell from the path
        return False

    if dfs(start[0], start[1]):
        return path
    else:
        return None

# 3. Topological Sort
def topological_sort(graph):
    """
    Performs topological sort on a directed acyclic graph (DAG) using DFS.

    Args:
        graph: A dictionary representing the graph as an adjacency list.

    Returns:
        A list of nodes in topological order, or None if the graph has a cycle.
    """
    visited = set()
    stack = []

    def dfs(node):
        """
        DFS function to traverse the graph.
        """
        if node in visited:
            return  # Already visited

        visited.add(node)
        for neighbor in graph.get(node, []):
            dfs(neighbor)
        stack.insert(0, node)  # Prepend the node to the result list

    for node in graph:
        dfs(node)

    # Check for cycles (Improved):
    if len(visited) != len(graph):
        return None # Cycle exists.  Not a DAG.

    return stack

# 4. Cycle Detection
def detect_cycle(graph):
    """
    Detects cycles in a directed graph using DFS.

    Args:
        graph: A dictionary representing the graph as an adjacency list.

    Returns:
        True if the graph contains a cycle, False otherwise.
    """
    visited = set()
    recursion_stack = set()  # Keep track of nodes in the current recursion stack

    def dfs(node):
        """
        DFS function to traverse the graph and detect cycles.
        """
        visited.add(node)
        recursion_stack.add(node)

        for neighbor in graph.get(node, []):
            if neighbor in recursion_stack:
                return True  # Cycle detected
            if neighbor not in visited:
                if dfs(neighbor):
                    return True

        recursion_stack.remove(node)  # Remove node from recursion stack after processing
        return False

    for node in graph:
        if node not in visited:
            if dfs(node):
                return True
    return False

# 5. Connected Components
def find_connected_components(graph):
    """
    Finds the connected components in an undirected graph using DFS.

    Args:
        graph: A dictionary representing the graph as an adjacency list.

    Returns:
        A list of sets, where each set represents a connected component.
    """
    visited = set()
    connected_components = []

    def dfs(node, component):
        """
        DFS function to traverse the graph and build a connected component.
        """
        visited.add(node)
        component.add(node)
        for neighbor in graph.get(node, []):
            if neighbor not in visited:
                dfs(neighbor, component)

    for node in graph:
        if node not in visited:
            component = set()
            dfs(node, component)
            connected_components.append(component)
    return connected_components

# 6. Sudoku Solver
def solve_sudoku(board):
    """
    Solves a Sudoku puzzle using DFS (backtracking).

    Args:
        board: A 9x9 2D list representing the Sudoku board.  0 represents empty cells.

    Returns:
        True if the Sudoku puzzle is solved, False otherwise.  The board is modified in place.
    """

    def find_empty_cell():
        """
        Finds the next empty cell (0) in the board.
        Returns:
            A tuple (row, col) of the empty cell's coordinates, or None if no empty cell exists.
        """
        for r in range(9):
            for c in range(9):
                if board[r][c] == 0:
                    return r, c
        return None

    def is_valid(r, c, num):
        """
        Checks if placing a number is valid in the given cell.

        Args:
            r: The row index.
            c: The column index.
            num: The number to place (1-9).

        Returns:
            True if the number is valid, False otherwise.
        """
        # Check row and column
        for i in range(9):
            if board[r][i] == num or board[i][c] == num:
                return False

        # Check 3x3 subgrid
        start_r = (r // 3) * 3
        start_c = (c // 3) * 3
        for i in range(3):
            for j in range(3):
                if board[start_r + i][start_c + j] == num:
                    return False
        return True

    def dfs():
        """
        DFS function to solve the Sudoku puzzle.
        """
        empty_cell = find_empty_cell()
        if not empty_cell:
            return True  # No empty cells left, puzzle is solved

        r, c = empty_cell
        for num in range(1, 10):
            if is_valid(r, c, num):
                board[r][c] = num
                if dfs():  # Recursively try to solve the rest of the board
                    return True
                board[r][c] = 0  # Backtrack: reset the cell if the solution doesn't work
        return False  # No valid number found for this cell

    return dfs()

def print_sudoku_board(board):
    """Prints the Sudoku board in a user-friendly format."""
    for r in range(9):
        if r % 3 == 0 and r != 0:
            print("-" * 21)  # Separator lines
        for c in range(9):
            if c % 3 == 0 and c != 0:
                print("|", end=" ")  # Separator bars
            print(board[r][c], end=" ")
        print()

if __name__ == "__main__":
    # 1. Web Crawler Example
    start_time = time.time()
    visited_urls = web_crawler("https://www.example.com", max_depth=2)
    end_time = time.time()
    print("\n[Web Crawler] Visited URLs:")
    for url in visited_urls:
        print(url)
    print(f"[Web Crawler] Time taken: {end_time - start_time:.4f} seconds")

    # 2. Maze Solving Example
    maze = [
        [0, 1, 1, 1, 1, 1, 1, 1],
        [0, 0, 0, 0, 0, 0, 0, 1],
        [1, 1, 1, 1, 1, 1, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 0],
        [1, 0, 1, 1, 0, 1, 1, 1],
        [1, 0, 0, 1, 0, 0, 0, 1],
        [1, 1, 0, 1, 1, 1, 0, 1],
        [1, 1, 0, 0, 0, 0, 0, 0],
    ]
    start = (0, 0)
    end = (7, 7)
    start_time = time.time()
    path = solve_maze_dfs(maze, start, end)
    end_time = time.time()
    print("\n[Maze Solving] Maze:")
    for row in maze:
        print(row)
    print(f"[Maze Solving] Path from {start} to {end}: {path}")
    print(f"[Maze Solving] Time taken: {end_time - start_time:.4f} seconds")

    # 3. Topological Sort Example
    graph = {
        'A': ['B', 'C'],
        'B': ['D', 'E'],
        'C': ['F'],
        'D': ['E'],
        'E': ['F'],
        'F': []
    }
    start_time = time.time()
    sorted_nodes = topological_sort(graph)
    end_time = time.time()
    print("\n[Topological Sort] Graph:")
    for node in graph:
        print(f"{node}: {graph[node]}")
    print(f"[Topological Sort] Topological ordering: {sorted_nodes}")
    print(f"[Topological Sort] Time taken: {end_time - start_time:.4f} seconds")

    # Example with a graph containing a cycle (for demonstration):
    graph_with_cycle = {
        'A': ['B'],
        'B': ['C'],
        'C': ['A']  # Cycle: C -> A
    }
    start_time = time.time()
    result_with_cycle = topological_sort(graph_with_cycle)
    end_time = time.time()
    print("\n[Topological Sort] Graph with cycle:")
    for node in graph_with_cycle:
        print(f"{node}: {graph_with_cycle[node]}")
    if result_with_cycle is None:
        print("[Topological Sort] Graph contains a cycle: Cannot perform topological sort.")
    else:
        print(f"[Topological Sort] Topological ordering: {result_with_cycle}")
    print(f"[Topological Sort] Time taken: {end_time - start_time:.4f} seconds")

    # 4. Cycle Detection Example
    graph_no_cycle = {
        'A': ['B', 'C'],
        'B': ['D', 'E'],
        'C': ['F'],
        'D': ['E'],
        'E': ['F'],
        'F': []
    }

    graph_with_cycle = {
        'A': ['B'],
        'B': ['C'],
        'C': ['A']  # Cycle: C -> A
    }
    start_time = time.time()
    has_cycle1 = detect_cycle(graph_no_cycle)
    has_cycle2 = detect_cycle(graph_with_cycle)
    end_time = time.time()
    print("\n[Cycle Detection] Graph without cycle:")
    for node in graph_no_cycle:
        print(f"{node}: {graph_no_cycle[node]}")
    print(f"[Cycle Detection] Has cycle: {has_cycle1}")  # False
    print("\n[Cycle Detection] Graph with cycle:")
    for node in graph_with_cycle:
        print(f"{node}: {graph_with_cycle[node]}")
    print(f"[Cycle Detection] Has cycle: {has_cycle2}")  # True
    print(f"[Cycle Detection] Time taken: {end_time - start_time:.4f} seconds")

    # 5. Connected Components Example
    graph = {
        0: [1, 2],
        1: [0, 2],
        2: [0, 1],
        3: [4],
        4: [3, 5],
        5: [4],
        6: [7, 8],
        7: [6, 8],
        8: [6, 7]
    }
    start_time = time.time()
    components = find_connected_components(graph)
    end_time = time.time()
    print("\n[Connected Components] Graph:")
    for node in graph:
        print(f"{node}: {graph[node]}")
    print(f"[Connected Components] Connected components: {components}")
    print(f"[Connected Components] Time taken: {end_time - start_time:.4f} seconds")

    # 6. Sudoku Solver Example
    # 0 represents empty cells
    sudoku_board = [
        [5, 3, 0, 0, 7, 0, 0, 0, 0],
        [6, 0, 0, 1, 9, 5, 0, 0, 0],
        [0, 9, 8, 0, 0, 0, 0, 6, 0],
        [8, 0, 0, 0, 6, 0, 0, 0, 3],
        [4, 0, 0, 8, 0, 3, 0, 0, 1],
        [7, 0, 0, 0, 2, 0, 0, 0, 6],
        [0, 6, 0, 0, 0, 0, 2, 8, 0],
        [0, 0, 0, 4, 1, 9, 0, 0, 5],
        [0, 0, 0, 0, 8, 0, 0, 7, 9]
    ]
    print("\n[Sudoku Solver] Initial Sudoku Board:")
    print_sudoku_board(sudoku_board)
    start_time = time.time()
    if solve_sudoku(sudoku_board):
        end_time = time.time()
        print("\n[Sudoku Solver] Solved Sudoku Board:")
        print_sudoku_board(sudoku_board)
        print(f"[Sudoku Solver] Time taken: {end_time - start_time:.4f} seconds")
    else:
        end_time = time.time()
        print("[Sudoku Solver] No solution exists.")
        print(f"[Sudoku Solver] Time taken: {end_time - start_time:.4f} seconds")

File: JavaScript/8-Graphs(BFS-DFS)/Algorithms/1-Depth-FirstSearch(DFS)/Depth-FirstSearch(DFS)Pro.js
// Depth-First Search (DFS) in JavaScript - 5 Approaches

// 1. Basic DFS Traversal (Graph Representation with Adjacency List)
//    - Real-world application: Finding connected components in a social network.

function dfsBasic(graph, startNode, visited = new Set()) {
  console.log("DFS Basic - Starting at node:", startNode);
  visited.add(startNode);
  console.log("DFS Basic - Visited node:", startNode);

  for (const neighbor of graph[startNode] || []) {
    if (!visited.has(neighbor)) {
      dfsBasic(graph, neighbor, visited);
    }
  }
}

const graphBasic = {
  'A': ['B', 'C'],
  'B': ['D', 'E'],
  'C': ['F'],
  'D': [],
  'E': ['F'],
  'F': []
};

console.log("1. Basic DFS Traversal");
dfsBasic(graphBasic, 'A');
console.log("------------------------------------------------------");

// 2. DFS to Find a Path (Path Existence)
//    - Real-world application:  Checking if a route exists between two locations in a navigation system.
function dfsFindPath(graph, startNode, endNode, visited = new Set()) {
    console.log(`DFS Path - Visiting node: ${startNode}, looking for ${endNode}`);
    if (startNode === endNode) {
        console.log(`DFS Path - Found path to: ${endNode}`);
        return true;
    }
    visited.add(startNode);

    for (const neighbor of graph[startNode] || []) {
        if (!visited.has(neighbor)) {
            if (dfsFindPath(graph, neighbor, endNode, visited)) {
                return true; // Path found!  Propagate the 'true' back up the call stack.
            }
        }
    }
    console.log(`DFS Path - No path from ${startNode} to ${endNode}`);
    return false; // No path found from this node
}

const graphPath = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': ['G'],
    'E': ['F'],
    'F': [],
    'G': []
};

console.log("2. DFS to Find a Path");
console.log("Path from A to F:", dfsFindPath(graphPath, 'A', 'F')); // true
console.log("Path from A to G:", dfsFindPath(graphPath, 'A', 'G'));
console.log("Path from C to B:", dfsFindPath(graphPath, 'C', 'B'));  //false
console.log("------------------------------------------------------");

// 3. DFS for Topological Sort (Directed Acyclic Graph)
//    - Real-world application: Task scheduling or dependency resolution in a build system.
function dfsTopologicalSort(graph, startNode, visited = new Set(), result = []) {
  console.log("DFS Topological - Visiting:", startNode);
  visited.add(startNode);

  for (const neighbor of graph[startNode] || []) {
    if (!visited.has(neighbor)) {
      dfsTopologicalSort(graph, neighbor, visited, result);
    }
  }
  result.unshift(startNode); // Add to the *beginning* of the result
  console.log("DFS Topological - Adding to result:", startNode, "Result:", result);
  return result;
}

function topologicalSort(graph) {
    const visited = new Set();
    const result = [];
    for (const node in graph) {
        if (!visited.has(node)) {
            dfsTopologicalSort(graph, node, visited, result);
        }
    }
    return result;
}
const graphTopological = {
  'A': ['C'],
  'B': ['C', 'D'],
  'C': ['E'],
  'D': ['F'],
  'E': ['F'],
  'F': []
};

console.log("3. DFS for Topological Sort");
console.log("Topological Sort:", topologicalSort(graphTopological));
console.log("------------------------------------------------------");

// 4. DFS to Detect Cycles in a Directed Graph
//     - Real-world application: Detecting circular dependencies in software projects.
function dfsCycleDetection(graph, startNode, visited = new Set(), recursionStack = new Set()) {
    console.log(`DFS Cycle - Visiting: ${startNode}, Visited: ${[...visited]}, Recursion Stack: ${[...recursionStack]}`);
    visited.add(startNode);
    recursionStack.add(startNode);

    for (const neighbor of graph[startNode] || []) {
        if (recursionStack.has(neighbor)) {
            console.log(`DFS Cycle - Cycle detected at: ${neighbor}`);
            return true; // Cycle detected!
        }
        if (!visited.has(neighbor)) {
            if (dfsCycleDetection(graph, neighbor, visited, recursionStack)) {
                return true; // Cycle detected in recursive call
            }
        }
    }

    recursionStack.delete(startNode); // Remove from recursion stack when finished
    console.log(`DFS Cycle - Finished with: ${startNode}, Recursion Stack: ${[...recursionStack]}`);
    return false; // No cycle found
}

function detectCycle(graph) {
    const visited = new Set();
    for (const node in graph) {
        if (!visited.has(node)) {
            if (dfsCycleDetection(graph, node, visited, new Set())) {
                return true; // Cycle found
            }
        }
    }
    return false; // No cycles
}

const graphCycle1 = {
    'A': ['B', 'C'],
    'B': ['C', 'D'],
    'C': ['D'],
    'D': ['A']  // Cycle: D -> A -> B -> C -> D
};

const graphCycle2 = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [],
    'E': ['F'],
    'F': []
};

console.log("4. DFS to Detect Cycles in a Directed Graph");
console.log("Cycle in graphCycle1:", detectCycle(graphCycle1)); // true
console.log("Cycle in graphCycle2:", detectCycle(graphCycle2)); // false
console.log("------------------------------------------------------");

// 5. DFS for Maze Solving
//    - Real-world application: Pathfinding in games or robotics.

function dfsMazeSolver(maze, start, end, path = [], visited = new Set()) {
  const rows = maze.length;
  const cols = maze[0].length;

  function isValid(r, c) {
    return r >= 0 && r < rows && c >= 0 && c < cols && maze[r][c] === 0; // 0 is path
  }

  function toKey(r, c) {
      return `${r},${c}`;
  }
  console.log(`DFS Maze - Visiting: ${start[0]},${start[1]}  End: ${end[0]},${end[1]}`);
  if (start[0] === end[0] && start[1] === end[1]) {
    path.push(start);
    console.log("DFS Maze - Found Path:", path);
    return true;
  }

  const [r, c] = start;
  visited.add(toKey(r,c));
  path.push(start);

  const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]; // Right, Down, Left, Up

  for (const [dr, dc] of directions) {
    const newR = r + dr;
    const newC = c + dc;

    if (isValid(newR, newC) && !visited.has(toKey(newR,newC))) {
      if (dfsMazeSolver(maze, [newR, newC], end, path, visited)) {
        return true;
      }
    }
  }

  path.pop(); // Backtrack
  console.log(`DFS Maze - Backtracking from: ${r},${c}`);
  return false;
}

const maze = [
  [0, 1, 0, 0, 0],
  [0, 1, 0, 1, 0],
  [0, 0, 0, 0, 0],
  [0, 1, 1, 1, 1],
  [0, 0, 0, 0, 0]
]; // 0: path, 1: wall

const start = [0, 0];
const end = [4, 4];

console.log("5. DFS for Maze Solving");
const pathMaze = [];
dfsMazeSolver(maze, start, end, pathMaze);
if (pathMaze.length > 0){
  console.log("Maze Solved! Path:", pathMaze);
}
else{
  console.log("No path found");
}

console.log("------------------------------------------------------");
